Exercises/chapter7/e1-a-robot.js

```diff
@@ -0,0 +1,192 @@
+      }
+    }
+  }
+}
+
+function randomRobot(state) {
+  return {direction: randomPick(roadGraph[state.place])};
+}
+function routeRobot(state, memory) {
+  const mailRoute = [
+    "Alice's House", "Cabin", "Alice's House", "Bob's House",
+    "Town Hall", "Daria's House", "Ernie's House",
+    "Grete's House", "Shop", "Grete's House", "Farm",
+    "Marketplace", "Post Office"
+  ];
+
+  if (memory.length === 0) {
+    memory = mailRoute;
+  }
+  return {direction: memory[0], memory: memory.slice(1)};
+}
+function goalOrientedRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    let parcel = parcels[0];
+    if (parcel.place != place) {
+      route = findRoute(roadGraph, place, parcel.place);
+    } else {
+      route = findRoute(roadGraph, place, parcel.address);
+    }
+  }
+  return {direction: route[0], memory: route.slice(1)};
+}
+
+/**
+  This robot moves towards the closest parcel until all parcels are collected.
+  It then visits the closest address of each undelivered parcel until all are
+  delivered.
+*/
+function closestParcelsRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    if (parcels.some(p => p.place != place)) {
+      let closestParcel = parcels
+        .filter(p => p.place !== place)
+        .map(parcel => {
+          route = findRoute(roadGraph, place, parcel.place);
+          return route
+        })
+        .reduce((r1, r2) => r1.length <= r2 ? r1 : r2);
+      route = closestParcel;
+    } else {
+      let closestAddress = parcels.map(parcel => {
+        route = findRoute(roadGraph, place, parcel.address);
+        return route
+      }).reduce((r1, r2) => r1.length <= r2 ? r1 : r2);
+      route = closestAddress;
+    }
+  }
+  return {direction: route[0], memory: route.slice(1)};
+}
+
+/**
+  This robot moves towards the closer of the closest parcel and closest address
+  of collected parcels until all parcels are delivered.
+*/
+function closestParcelOrAddressRobot({place, parcels}, route) {
+  if (route.length === 0) {
+    route = parcels
+      .map(parcel => {
+        if (parcel.place != place) {
+          return findRoute(roadGraph, place, parcel.place);
+        } else {
+          return findRoute(roadGraph, place, parcel.address);
+        }
+      })
+      .reduce((r1, r2) => r1.length <= r2 ? r1 : r2);
+  }
+  return {direction: route[0], memory: route.slice(1)};
+}
+
+function compareRobots(robot1, memory1, robot2, memory2) {
+  let robot1turns = 0;
+  let robot2turns = 0;
+  for (let i = 0; i < 100; i++) {
+    let task = VillageState.random();
+    robot1turns += runRobot(task, robot1, memory1);
+    robot2turns += runRobot(task, robot2, memory2);
+  }
+  let robot1AvgTurns = Math.round(robot1turns/100);
+  let robot2AvgTurns = Math.round(robot2turns/100);
+  console.log(`Robot1 took an average of ${robot1AvgTurns} turns.`);
+  console.log(`Robot2 took an average of ${robot2AvgTurns} turns.`);
+}
+
+compareRobots(routeRobot, [], goalOrientedRobot, []);
```

Comment:
It would be interesting to see a robot that picks up the closest package and then delivers it, or that picks up a random package and delivers it as a point of comparison.